### ISSUE TITLE
Provide a K-V Store API in LS-Core

### ISSUE THREAD

**Issue by original-brownbear** [elastic/logstash#8374]
@suyograo let me know if I got this right :)

#### We want

Generic K-V store API usable by plugins and in LS core, backed by a configurable store (in step 1 this should be ES). 

#### How about this?

* Configure state store in `logstash.yml` (es host etc.)
* If not configured use temp file that is only valid for the lifetime of the LS process
* Provide Java static API that allows you to get an isolated store by name like `LsStateStore.getOrCreate("NameOfStatestore")` that will then be backed by the configured backend from in the `logstash.yml`
* Have `LsStateStore` provide `get()`, `set()`, `delete()` for any `String` key and `String` value

How does that sound?

**Comment by jordansissel** [comment]
I'd prefer a typed and serializable value instead of a string.

For example, 
* File input stores a map (file identifier -> file state) where the file state is a structure with enough detail to resume reading when restarted.
* The S3 input has a similar problem to the file input
* JDBC input needs to store the "sql_last_value" where the type (integer, timestamp, etc) is defined by the user's database query.

**Comment by jordansissel** [comment]
> If not configured use temp file that is only valid for the lifetime of the LS process

This would break some expectations, for example, the file input and jdbc inputs would lose state on restart if this temporary-file approach is used. I believe the default should be a predictable path (relative to path.data, perhaps) on the local filesystem.

**Comment by jordansissel** [comment]
> Provide Java static API 

This makes it hard to do isolation and dependency injection (injecting specific settings for a given test) during tests. Does it need to be static? What's the benefit of static vs providing the service to the plugin directly?

**Comment by original-brownbear** [comment]
@jordansissel 

> What's the benefit of static vs providing the service to the plugin directly?

No other benefit than not having to change the plugin API :)
Obviously, you could pass this down during `register` or something but it's more effort. That said, you can always use some random string as the store identifier and it's fine for tests imo.

**Comment by original-brownbear** [comment]
@jordansissel sure, so just any `Serializable`as value? That should work fine.

**Comment by suyograo** [comment]
> I believe the default should be a predictable path (relative to path.data, perhaps) on the local filesystem.

Agreed. Today for file input we use `<path.data>/plugins/inputs/file`. See https://github.com/logstash-plugins/logstash-input-file/blob/master/lib/logstash/inputs/file.rb#L117

**Comment by suyograo** [comment]
As a fallback, if state store is not configured, we should use the one in `path.data`
